### ISSUE TITLE
Hooking Callback not Executed

### ISSUE THREAD

**Issue by oleavr** [frida/frida-gum#187]
(Continuation of frida/frida#215.)

As described by @AndyXan:

I'm reopening the Issue as stated. I retested the example with the newest Frida version. The first test was on a Nexus Tablet running 6.0.1. I Retested the same with an ARM-32 Emulator, Running 4.4.2 resulting in the exact same behaviour. Still i'm new to Frida - might be some stupid mistake of mine. 

Native Code from SubstrateTestApp: [targetApp1-debug.apk.zip](https://github.com/frida/frida/files/714005/targetApp1-debug.apk.zip)

```
int getAge(void)
{
    LOGI("[i] Verbose - getAge located at %p\n", &getAge);
    return 21;
}

jstring
Java_io_koz_targetApp1_targetApp1_doThings( JNIEnv* env,
                                                  jobject thiz )
{
    int r = arc4random() % 10000;
    LOGI("[+] John Smith is %d years old.\n", getAge());
    LOGI("[+] The totally reliable random seed is: %d\n\n", r);
    return 0;
}
```

Hooking the JNI Call works perfectly fine. The Message appear in the console.

```
package_name = "io.koz.targetApp1"
...  
def hookGetAge():
    get_age_hook_code = """
       Interceptor.attach(Module.findExportByName ("libtargetLib.so","getAge"), {
            onEnter: function (args) {
                send("[*] Instrumentando...");
                }
    });
    """
    return get_age_hook_code
...
script = session.create_script(hookGetAge())
script.on('message', on_message)
script.load()
sys.stdin.read()
```

However, trying to hook the second native function GetAge() which gets executed according to logcat does not trigger the callback at all. The Symbol getAge gets found by findExportByName and can be resolved (changign getAge to getAgeeeee leads to exception). So it is not a problem with the symbol. Inserting a correct runtime address (ptr(..)) also does not work. 

```
def hookGetAge():
    get_age_hook_code = """
       Interceptor.attach(Module.findExportByName ("libtargetLib.so","getAge"), {
            onEnter: function (args) {
                send("[*] Instrumentando...");
                }
    });
    """
    return get_age_hook_code
```

Environment Ubuntu 13.01, Device: Rooted Tablet Nexus, Android 6.0 including SuperSu and so on. Rechecked with Emulator Running 4.4.2. 

Hope somebody can help me here!
Greetings

**Comment by AndyXan** [comment]
To narrow it down a bit:

Hooking the `arc4random() `function also works with `Interceptor.attach(Module.findExportByName ("libc.so","arc4random")`

Also, the `findExportByName ("libtargetLib.so","getAge")` returns the correct value. When returning `send(Module.findExportByName ("libtargetLib.so","getAge"))` inside the Jni call where the hook works the value returned equals the value displayed in the logcat output from  `LOGI("[i] Verbose - getAge located at %p\n", &getAge);
`

From an IDA-Point of View its, getAge() gets called ADD, Register, PC instead of BL/BLX. 
```
.text:00000F7C                 LDR             R3, [R0] ; getAge
.text:00000F7E                 ADD             R2, PC  ; "[i] Verbose - getAge located at %p\n"
```

So basically.. i have no idea :)

**Comment by AndyXan** [comment]
And another thing which might be the root cause of all this.

When manipulating the original targetLib.c to:

```
int getMoreAge()
{
    LOGI("[i] Verbose - getMoreAge located at %p\n", &getMoreAge);
    LOGI("[i] Verbose - making more system calls and stuff");
    return 30;
}

int getAge()
{
    LOGI("[i] Verbose - getAge located at %p\n", &getAge);
    LOGI("[i] Verbose - making more system calls and stuff");
    int f = getMoreAge();
```

Basically adding another call in getAge() to getMoreAge(), the hook 

```
def hookGetAge():
    get_age_hook_code = """
       Interceptor.attach(Module.findExportByName ("libtargetLib.so","getAge"), {
            onEnter: function (args) {
                send("[*] Instrumentando...");
                }
    });
    """
    return get_age_hook_code
```

is working. However, the hook for getMoreAge() does not work. 

So i assume it has something to do with a function being the last function in call chain.

Visualized:

Java_io_koz_targetApp1_targetApp1_doThings -> getAge()
doThings hook works. getAge hook does not work.

Java_io_koz_targetApp1_targetApp1_doThings -> getAge() -> getMoreAge()
doThings hook works. GetAge hook works. GetMoreAge() hook does not work.
